polkadot-test-service/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		execution_strategies: ExecutionStrategies {
+			syncing: sc_client_api::ExecutionStrategy::NativeWhenPossible,
+			importing: sc_client_api::ExecutionStrategy::NativeWhenPossible,
+			block_construction: sc_client_api::ExecutionStrategy::NativeWhenPossible,
+			offchain_worker: sc_client_api::ExecutionStrategy::NativeWhenPossible,
+			other: sc_client_api::ExecutionStrategy::NativeWhenPossible,
+		},
+		rpc_http: None,
+		rpc_ws: None,
+		rpc_ipc: None,
+		rpc_ws_max_connections: None,
+		rpc_cors: None,
+		rpc_methods: Default::default(),
+		prometheus_config: None,
+		telemetry_endpoints: None,
+		telemetry_external_transport: None,
+		default_heap_pages: None,
+		offchain_worker: Default::default(),
+		force_authoring: false,
+		disable_grandpa: false,
+		dev_key_seed: Some(key_seed),
+		tracing_targets: None,
+		tracing_receiver: Default::default(),
+		max_runtime_instances: 8,
+		announce_block: true,
+		base_path: Some(base_path),
+		informant_output_format,
+	})
+}
+
+/// Run a Polkadot test node using the Polkadot test runtime. The node will be using an in-memory socket, therefore you
+/// need to provide boot nodes if you want it to be connected to other nodes. The `storage_update_func` can be used to
+/// make adjustements to the runtime before the node starts.
+pub fn run_test_node(
+	task_executor: TaskExecutor,
+	key: Sr25519Keyring,
+	storage_update_func: impl Fn(),
+	boot_nodes: Vec<MultiaddrWithPeerId>,
+) -> Result<
+	PolkadotTestNode<
+		TaskManager,
+		impl PolkadotClient<Block, TFullBackend<Block>, polkadot_test_runtime::RuntimeApi>,
+	>,
+	ServiceError,
+> {
+	let config = node_config(storage_update_func, task_executor, key, boot_nodes)?;
+	let multiaddr = config.network.listen_addresses[0].clone();
+	let authority_discovery_enabled = false;
+	let (task_manager, client, handles, network, rpc_handlers) =
+		polkadot_test_new_full(config, None, None, authority_discovery_enabled, 6000)?;
+
+	let peer_id = network.local_peer_id().clone();
+	let addr = MultiaddrWithPeerId { multiaddr, peer_id };
+
+	let node = PolkadotTestNode {
+		task_manager,
+		client,
+		handles,
+		addr,
+		rpc_handlers,
+	};
+
+	Ok(node)
+}
+
+/// A Polkadot test node instance used for testing.
+pub struct PolkadotTestNode<S, C> {
+	/// TaskManager's instance.
+	pub task_manager: S,
+	/// Client's instance.
+	pub client: Arc<C>,
+	/// Node's handles.
+	pub handles: FullNodeHandles,
+	/// The `MultiaddrWithPeerId` to this node. This is useful if you want to pass it as "boot node" to other nodes.
+	pub addr: MultiaddrWithPeerId,
+	/// RPCHandlers to make RPC queries.
+	pub rpc_handlers: Arc<RpcHandlers>,
+}
+
+impl<S, C> PolkadotTestNode<S, C>
+where
+	C: HeaderBackend<Block>,
+{
+	/// Send a transaction through RPCHandlers to call a function.
+	pub async fn call_function(
+		&self,
+		function: polkadot_test_runtime::Call,
+		caller: Sr25519Keyring,
+	) -> (
+		Option<String>,
```

Comment:
Could we not make this function return a result or an error if the sending failed? Instead of all these 3 values which not say anything what they are for.

Reply:
We can but there are some annoyances:

1. I will need to parse the JSON returned and add serde_json to the direct dependencies
2. we still need to return the raw output whatever that is (String or serde_json::Value)
3. the other 2 objects of the tuple should be returned but we can definitely drop them if you want (we just lose the callback mechanism)